### DISCUSSION TITLE
Query about Arabic cursivity property

### DISCUSSION THREAD

**Discussion by sid-wr** [harfbuzz/harfbuzz#4527] [General] [answered]
Hi @behdad @khaledhosny 

We are trying to do justification for Arabic text on our side. For proper justification handling of arabic text we require to know which form(isolated,initial,medial,final) is returned by harfbuzz. Currently hb_shape.exe shows this output as is seen in below example( we can se .init,.medi,.fina etc)

```
hb-shape Amiri-Regular.ttf "به بـه" --show-flags --no-positions --no-clusters
```

Output:
```
[heh-ar.fina#3|kashida-ar.1#3|beh-ar.init|space|heh-ar.fina#3|beh-ar.init]
```
We would like to know if we can get this information from hb_buffer or any flags in glyph_info so that we can properly set the cursivity connections between these glyphs

Thanks

**Comment by khaledhosny** [comment]
I don’t understand what you are trying to do.

**Comment by sid-wr** [comment]
Hi @khaledhosny,

I am trying to do full Arabic justification on my end after using HarfBuzz.
For this I need the cursivity info of the glyphs returned by HarfBuzz.

I have some custom justification logic which treats isolated glyphs as separate units of text & stretches them away during justification.

As the Arabic script shaping is complex one according to the fact that we can't assume n number of unicodes map to n number of glyphs always. So it is m:n mapping (m=n or m≠n) unlike n:n mapping in Latin scripts.

From HarfBuzz, I only have the final glyphs after shaping.
For the example input `تقنين` assuming RTL direction:
- (**Currently**) if the **second last glyph has form=final** & **last glyph has form=isolated** (which are coming incorrect due to my logic), then the text gets displayed as follows:
![Current behaviour](https://github.com/harfbuzz/harfbuzz/assets/136314541/4be02de5-1c58-4cf7-b9d5-6ee0cae2e503)
- (**Desirable**) if the **second last glyph has form=medial** & **last glyph has form=final**, then the text gets displayed as follows:
![Desired behaviour](https://github.com/harfbuzz/harfbuzz/assets/136314541/537627ba-a21f-491d-b2d6-5434f55e673b)


That's why I was asking if HarfBuzz provides any information from hb_buffer or any flags in glyph_info so that we can properly set the cursivity connections between these glyphs

**Reply by khaledhosny** [reply, marked as answer]
I still don’t understand what you are trying to do, since there shouldn’t be any increase of spacing inside Arabic words as it would make the text unreadable. If you need to increase spacing, it should be applied only to word spaces (i.e. space and other similar characters).

Elongating Arabic words for justification is done by inserting kashida at certain places. Where to insert Kashida has a lot of variation, but a relatively widely used algorithm is the one implemented by Internet Explorer (documented [here](https://web.archive.org/web/20180115010153/http://www.microsoft.com/middleeast/msdn/JustifyingText-CSS.aspx)) and it mostly works by checking certain pairs of letters, so the joining behavior of these letters is assumed, except in the last rule, where you would nee to check [`ArabicShaping.txt`](https://unicode.org/Public/UNIDATA/ArabicShaping.txt) for the joining behavior of the code points and decide whether they are connected or not. You shouldn’t need any help from HarfBuzz here.

The `HB_GLYPH_FLAG_SAFE_TO_INSERT_TATWEEL` can then be used to filter-out kashida insertion positions that would break glyph connection (and this is relevant only if you are going to insert kashida glyphs without reshaping the text, if you insert kashida characters in the input string and shape the text again, then you don’t need this flag either).

**Comment by sid-wr** [comment]
Hi @khaledhosny @behdad

Can you please confirm whether HarfBuzz provides any such information from hb_buffer or any flags in glyph_info so that we can properly set the cursivity connections between these glyphs.
